File: cay_khung_nho_nhat.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Định nghĩa cấu trúc biểu diễn cạnh của đồ thị
struct Edge {
    int u, v, weight;
    Edge(int _u, int _v, int _weight) : u(_u), v(_v), weight(_weight) {}
};

// Định nghĩa cấu trúc biểu diễn đồ thị
class Graph {
private:
    int n; // Số đỉnh
    vector<vector<pair<int, int>>> adj; // Danh sách kề

public:
    Graph(int _n) : n(_n) {
        adj.resize(n + 1);
    }

    // Thêm cạnh vào đồ thị
    void addEdge(int u, int v, int weight) {
        adj[u].push_back({v, weight});
        adj[v].push_back({u, weight});
    }

    // Tìm cây khung nhỏ nhất bằng thuật toán Prim
    int findMinimumSpanningTree() {
        int totalWeight = 0;
        vector<bool> visited(n + 1, false);
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

        // Bắt đầu từ đỉnh 1
        pq.push({0, 1});

        while (!pq.empty()) {
            int u = pq.top().second;
            int weight = pq.top().first;
            pq.pop();

            // Nếu đỉnh này đã được thăm, bỏ qua
            if (visited[u]) continue;

            visited[u] = true;
            totalWeight += weight;

            for (const auto& neighbor : adj[u]) {
                int v = neighbor.first;
                int w = neighbor.second;
                if (!visited[v]) {
                    pq.push({w, v});
                }
            }
        }

        return totalWeight;
    }
};

int main() {
    int n, m;
    cin >> n >> m;

    Graph graph(n);

    for (int i = 0; i < m; ++i) {
        int u, v, weight;
        cin >> u >> v >> weight;
        graph.addEdge(u, v, weight);
    }

    int minSpanningTreeWeight = graph.findMinimumSpanningTree();

    cout << minSpanningTreeWeight << endl;

    return 0;
}

File: cay_khung_nho_nhat.py
import heapq

class Graph:
    def __init__(self, n):
        self.n = n
        self.adj = [[] for _ in range(n + 1)]

    def add_edge(self, u, v, weight):
        self.adj[u].append((v, weight))
        self.adj[v].append((u, weight))

    def find_minimum_spanning_tree(self):
        total_weight = 0
        visited = [False] * (self.n + 1)
        pq = [(0, 1)]

        while pq:
            weight, u = heapq.heappop(pq)

            if visited[u]:
                continue

            visited[u] = True
            total_weight += weight

            for neighbor in self.adj[u]:
                v, w = neighbor
                if not visited[v]:
                    heapq.heappush(pq, (w, v))

        return total_weight

if __name__ == "__main__":
    n, m = map(int, input().split())

    graph = Graph(n)

    for _ in range(m):
        u, v, weight = map(int, input().split())
        graph.add_edge(u, v, weight)

    min_spanning_tree_weight = graph.find_minimum_spanning_tree()

    print(min_spanning_tree_weight)
